source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -106,25 +111,73 @@ void InstanceImpl::ThreadLocalPool::onClusterRemoval(const std::string& cluster_
 
   // Treat cluster removal as a removal of all hosts. Close all connections and fail all pending
   // requests.
+  if (host_set_member_update_cb_handle_ != nullptr) {
+    host_set_member_update_cb_handle_->remove();
+    host_set_member_update_cb_handle_ = nullptr;
+  }
   while (!client_map_.empty()) {
     client_map_.begin()->second->redis_client_->close();
   }
+  while (!clients_to_drain_.empty()) {
+    (*clients_to_drain_.begin())->redis_client_->close();
+  }
 
   cluster_ = nullptr;
-  host_set_member_update_cb_handle_ = nullptr;
   host_address_map_.clear();
 }
 
+void InstanceImpl::ThreadLocalPool::onHostsAdded(
+    const std::vector<Upstream::HostSharedPtr>& hosts_added) {
+  for (const auto& host : hosts_added) {
+    std::string host_address = host->address()->asString();
+    // Insert new host into address map, possibly overwriting a previous host's entry.
+    host_address_map_[host_address] = host;
+    for (const auto& created_host : created_hosts_) {
```

Comment:
Is it possible to reuse the connections?  Otherwise, when the cluster scale out, we can potentially create a large number of new connections through redirect, then quickly throw those connections away at the same time during the next cluster update.

Reply:
it would take some refactoring... the underlying connection logic depends on a host. I looked at swapping out hosts, but decided against the extra work (and changes to common upstream objects), given the fact that we should know the cluster topology most of the time.

Reply:
sure ... I'll add a cluster-scoped counter for clients that are drained.